h/models/group.py

```diff
@@ -152,6 +169,50 @@ def _write_principal(group):
     }.get(group.writeable_by)
 
 
+class _GroupTypeFlags(object):
+    """Abstract base class for group type flags classes."""
+
+    def __eq__(self, other):
+        """Return True if other has the same access flags as this matcher."""
+        attrs = ('joinable_by', 'readable_by', 'writeable_by')
+        for attr in attrs:
+            self_attr = getattr(self, attr)
+            other_attr = getattr(other, attr, None)
+            if self_attr != other_attr:
+                return False
+        return True
+
+    def __ne__(self, other):
+        """Return True if other has different access flags than this matcher."""
+        return not self.__eq__(other)
+
+
+class OpenGroupTypeFlags(_GroupTypeFlags):
```

Comment:
Conceptually it looks like `OpenGroupTypeFlags` and `PrivateGroupTypeFlags` should be instances of `GroupTypeFlags` rather subclasses since they do not have additional properties or behaviour. Am I missing something?

Reply:
True, they could be one class, but then that class's `__init__()` method would have to take a `type` argument and would have to branch on type and raise if an invalid type were given, and users of the class would need to pass a correct type string when instantiating. Subclasses may be a nicer API as you just do `OpenGroupTypeFlags()` or `PrivateGroupTypeFlags()`, and there's no way to attempt to construct an `InvalidGroupTypeFlags()`.